charms/worker/k8s/src/token_distributor.py

```diff
@@ -24,10 +31,92 @@
 
 log = logging.getLogger(__name__)
 
-SECRET_ID = "{0}-secret-id"  # nosec
-
 UNIT_RE = re.compile(r"k8s(-worker)?/\d+")
 
+RELATION_NODE_NAME = "node-name"
+RELATION_JOINED = "joined"
+RELATION_CLUSTER_NAME = "cluster-name"
+RELATION_TOKEN_FAILURE = "token-failure"  # nosec
+RELATION_SECRET_ID = "{0}-secret-id"  # nosec
+
+
+class TokenFailure(BaseModel):
+    """Model for token failure."""
+
+    revision: int
+    error: str
+
+
+class TokenContent(BaseModel):
+    """Model for token data for the relation's secret.
+
+    When serialized, this will be a object with the following fields:
+        revision (str): The revision number of the token as a string
+        token (str): The token string, expose
+
+        whether serialized to json or to a dict -- always a dict[str,str]
+
+    When deserialized, this will be a object with the following fields:
+        revision (int): The revision number of the token as an int
+        token (SecretStr): The token string, exposed as a SecretStr for security
+    """
+
+    revision: int = Field(default=0)
+    token: SecretStr
+
+    @field_validator("revision", mode="before", json_schema_input_type=Union[int, str])
+    @classmethod
+    def as_int(cls, v: str):
+        """Convert revision to int."""
+        return int(v)
+
+    @field_serializer("revision", when_used="always")
+    def _revision_as_str(self, v: int) -> str:
+        """Convert revision to str for serialization."""
+        return str(v)
+
+    @field_serializer("token", when_used="always")
+    def _token_as_str(self, v: SecretStr) -> str:
+        """Expose token to str for serialization."""
+        return v.get_secret_value()
```

Comment:
this is the pile of serializer/deserializers to make sure that dict or json model dumps always create a `dict[str,str]` for juju secrets to use